### ISSUE TITLE
The NotNullWhen attribute is not respected for out parameters

### ISSUE THREAD

**Issue by catmeow72** [dotnet/roslyn#61073]
**Version Used**: 
4.2.0.0 from OmniSharp,
The version(s) used in these versions of the .NET SDK:
 - 6.0.202-1 from Microsoft's official Ubuntu focal package repositories (Tested on Elementary OS 6.1)
 - 6.0.202 on Alpine Linux

**Steps to Reproduce**:

1. Clone  [this repo](https://github.com/catmeow72/notnullwhen-bug-example)
2. Build the project and see three erroneous warnings appear due to the NotNullWhen attribute not being respected for out parameters

**Expected Behavior**:
There should not be possible null reference warnings because the if statements checking the return value of the function with the NotNullWhen attribute on the out parameter do not allow the reference to be null
Additionally, it should act the same whether the out parameters are used to check if they are null or the return value is used to check the possibility of a null value.

**Actual Behavior**:
Roslyn ignores the NotNullWhen attribute and generates warnings for these variables anyway

**Comment by CyrusNajmabadi** [comment]
Hey Zachary, any chance you can make a smaller repro? It just show what warnings you were getting and what code it was being reported on?

**Comment by catmeow72** [comment]
With a function like this: 
```cs
bool Example([NotNullWhen(true)] out string? output) {
  output = "Hello, world!";
  return true;
}
```
it will show a warning when using the output of the function that requires a non-null reference, even if checking the return value of the function, such as this:
```cs
bool isNotNull = Example(out var output);
if (isNotNull) {
  RequiresNonNullValue(output);
}
```

**Comment by CyrusNajmabadi** [comment]
This is my design @catmeow72   c# does not track the relationship between different locations like this. You need to do `if(Example(...))` for it to be understood

**Comment by CyrusNajmabadi** [comment]
Note: it you would like this to work, please open a discussion at dotnet/csharplang. Thanks!

**Comment by catmeow72** [comment]
Sorry, I just didn't think of just doing that until after I opened this issue.

**Comment by CyrusNajmabadi** [comment]
Not a problem :-)
